derive/src/lib.rs

```diff
@@ -104,20 +110,38 @@ fn generate_fields(fields: &FieldsList) -> Vec<TokenStream2> {
         .iter()
         .map(|f| {
             let (ty, ident) = (&f.ty, &f.ident);
+            let type_name = clean_type_string(&quote!(#ty).to_string());
 
             if let Some(i) = ident {
                 quote! {
-                    .field_of::<#ty>(stringify!(#i), stringify!(#ty))
+                    .field_of::<#ty>(stringify!(#i), #type_name)
                 }
             } else {
                 quote! {
-                    .field_of::<#ty>(stringify!(#ty))
+                    .field_of::<#ty>(#type_name)
                 }
             }
         })
         .collect()
 }
 
+fn clean_type_string(input: &str) -> String {
+    input
+        .replace(" ::", "::")
+        .replace(":: ", "::")
+        .replace(" ,", ",")
+        .replace(" ;", ";")
+        .replace(" [", "[")
+        .replace("[ ", "[")
+        .replace(" ]", "]")
+        .replace(" (", "(")
+        .replace("( ", "(")
+        .replace(" )", ")")
+        .replace(" <", "<")
+        .replace("< ", "<")
+        .replace(" >", ">")
```

Comment:
:laughing:

Reply:
dq: why wouldn't `.replace(" ", "");` work here? Is space ever significant?

Reply:
yes, e.g. something like `&mut Foo` would become `&mutFoo`

Reply:
🤦

Reply:
I'd love to claim credit for this beauty but I stole it from https://github.com/paritytech/substrate/blob/master/frame/support/procedural/tools/src/lib.rs#L96